bredis/AsyncConnection.hpp

```diff
@@ -51,24 +53,75 @@ template <typename S> class AsyncConnection {
     std::mutex tx_queue_mutex_;
     std::mutex rx_queue_mutex_;
 
+    // I think that queuing should be the responsibility
+    // of a higher level interface and not this class.
+    // Instead, this class should just allow for one
+    // active asynchronous Redis read and one active
+    // asynchronous Redis write, with undefined behavior
+    // if two concurrent reads or two concurrent writes
+    // happen.
+    //
+    // Rationale: TCP/IP provides application-level flow
+    // control to match the rate of the receiver and the
+    // sender. In this implementation, the reader attempts
+    // to read as quickly as possible no matter how fast
+    // the consumer (in this case the command_callback_t)
+    // can process the data. This not only circumvents
+    // TCP/IP's natural application level flow control,
+    // but allows the size of the respective queues to
+    // grow without bounds. Furthermore, the caller is
+    // given no opportunity to provide "back pressure",
+    // nor is the caller able to detect the unbounded
+    // growth and impose its own policies on it.
+    // 
```

Comment:
Okay, I have studied your code a little more closely. Please tell me if I have it right. In this code:
https://github.com/basiliscos/cpp-bredis/blob/master/bredis/impl/async_connection.hpp#L42

You are atomically taking ownership of the entire write queue, and then in a single network call you are sending all of the requests that used to be in that queue?

If you send a ping command asynchronously in a loop, you should get the same result. This is because the Nagle algorithm will hold back sending of the TCP/IP packet until the kernel buffer has enough pings in it to form a complete packet. You just need to make sure that the Nagle algorithm can do its job. You shouldn't wait for the corresponding read to complete before sending the next ping. That's how pipelining works. So I think you can rewrite your test to accomplish this, without needing to change the implementation.